### PR TITLE
Cleaning house in MathUtils.

### LINKED ISSUE #3310: Merge GATKProtectedMathUtils into MathUtils

### LINKED ISSUE #3311: Redundant implementations of dotProduct()
We have more than one.

### LINKED ISSUE #4151: Wasteful MathUtils::sum methods
`MathUtils.sum(double[] array, int start, int stop)` copies a subarray and then sums the copy.  There is no need to make a copy.

### PR DESCRIPTION
**Pull request by davidbenjamin**
Closes #3310. Closes #3311. Closes #4151.

@samuelklee Could you review these Boy Scout activities?

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java [hunk 1]
```diff
@@ -44,12 +59,408 @@
...
+
+    /**
+     * Returns the smallest power of 2 that exceeds or equals {@code val}
+     * @param val input value
+     * @return power of 2 integer
+     * @throws IllegalArgumentException if an int overlow is encountered
+     */
+    public static int smallestPowerOfTwoGreaterThan(final int val)
+            throws IllegalArgumentException {
+        if (val > Integer.MAX_VALUE/2 || val < 0)
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: Use Utils.validateArg, clean up typos, and remove `throws` from signature.

File: src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java [hunk 2]
```diff
@@ -44,12 +59,408 @@
...
+     *      old grid = x     x     x
+     *      new grid = x   x   x   x
+     *      new data = 1   2   2   3
+     *
+     * @param data original array
+     * @param newLength length of the new array
+     * @return interpolated array
+     * @throws IllegalArgumentException if the input array is empty
+     */
+    public static double[] nearestNeighborUniform1DInterpolate(@Nonnull final double[] data, final int newLength)
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: Remove `throws` and perform null checks.

File: src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java [hunk 3]
```diff
@@ -44,12 +59,408 @@
...
+        }
+        if (points == 0) {
+            return new double[] {};
+        }
+        return IntStream.range(0, points).mapToDouble(n -> min + n * (max - min) / (points - 1)).toArray();
+    }
+
+    // given a list of options and a function for calculating their probabilities (these must sum to 1 over the whole list)
+    // randomly choose one option from the implied categorical distribution
+    public static <E> E randomSelect(final List<E> choices, final Function<E, Double> probabilityFunction, final RandomGenerator rng) {
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: Perform null checks.

File: src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java [hunk 4]
```diff
@@ -44,12 +59,408 @@
...
+        double sum = 0.0;
+        for (int i = 0; i < values.length; ++i) {
+            if (values[i] != Double.NEGATIVE_INFINITY) {
+                sum += Math.exp(values[i] - max);
+            }
+        }
+        return max + Math.log(sum);
+    }
+
+    public static double logSumExp(final Collection<Double> values) {
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: Perform null check.

File: src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java [hunk 5]
```diff
@@ -44,12 +59,408 @@
...
+     * Computes $\log(\sum_i e^{a_i})$ trying to avoid underflow issues by using the log-sum-exp trick.
+     *
+     * <p>
+     * This trick consists on shift all the log values by the maximum so that exponent values are
+     * much larger (close to 1) before they are summed. Then the result is shifted back down by
+     * the same amount in order to obtain the correct value.
+     * </p>
+     * @return any double value.
+     */
+    public static double logSumExp(final double ... values) {
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: Perform null check.

File: src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java [hunk 6]
```diff
@@ -44,12 +59,408 @@
...
+        Utils.validateArg(data.length > 0, "The input array is empty.");
+        Utils.validateArg(newLength >= 2, "The new length of the array must be >= 2");
+        final double fact = (double)(data.length - 1)/(newLength - 1);
+        return IntStream.range(0, newLength).mapToDouble(i -> data[(int) FastMath.floor(i*fact + 0.5)]).toArray();
+    }
+
+    /**
+     * Find the maximum difference between entries of two arrays.  This is useful for testing convergence, for example
+     */
+    public static double maxDifference(final List<Double> array1, final List<Double> array2) {
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: Add null checks here and below where necessary.

File: src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java [hunk 7]
```diff
@@ -44,12 +59,408 @@
...
     /**
      * Private constructor.  No instantiating this class!
      */
     private MathUtils() { }
 
+    /**
+     * Computes $\log(\sum_i e^{a_i})$ trying to avoid underflow issues by using the log-sum-exp trick.
+     *
+     * <p>
+     * This trick consists on shift all the log values by the maximum so that exponent values are
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java (hunk 5 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java (hunk 4 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java (hunk 3 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java [hunk 8]
```diff
@@ -44,12 +59,408 @@
...
+    public static double sum(final double[][][] array) {
+        Utils.nonNull(array);
+        double result = 0;
+        for (double[][] d: array) {
+            for (double[] da: d){
+                for (double daa: da) {
+                    result += daa;
+                }
+            }
+
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java (hunk 1 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: done done done

File: src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java (hunk 2 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: done done

File: src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java (hunk 6 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: done

**Comment by davidbenjamin** [comment]
Back to @samuelklee.

File: src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java (hunk 6 above)
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: I think there are a few more methods with `double[]` parameters below?

**Comment by samuelklee** [comment]
Looks good, just a few more missed null checks.  No need to send back!

**Pull request merged**